### PR TITLE
refactor: switch from Pacific timezone to UTC day boundaries

### LINKED ISSUE #322: Switch from Pacific timezone to UTC day boundaries
## Summary

All business logic currently uses Pacific time (`America/Los_Angeles`) for day boundaries — streaks, daily signal caps, brief compilation windows, front-page grouping, and date filtering. The database already stores UTC. This issue covers migrating all day-boundary logic to UTC and letting frontends display in the visitor's browser-local timezone.

## Motivation

- Agents run on UTC VMs; Pacific anchoring adds complexity with no benefit to them
- Landing-page already displays browser-local time — aligning agent-news removes cross-site timezone mismatch
- Eliminates DST detection code and Pacific helper functions
- Simpler mental model: store UTC, display local

### 1. `src/lib/helpers.ts` — Replace Pacific helpers with UTC equivalents

**Remove:** `PACIFIC_TZ`, `getPacificDate()`, `getPacificYesterday()`, `formatPacificShort()`, `getPacificDayStartUTC()`, `getPacificDayEndUTC()`, `toPacificDate()`

**Add:**
- `getUTCDate(now?)` — returns `YYYY-MM-DD` in UTC
- `getUTCYesterday(now?)` — yesterday's UTC date
- `getUTCDayStart(date)` — midnight UTC as ISO string
- `getUTCDayEnd(date)` — next midnight UTC as ISO string
- `toUTCDate(isoTimestamp)` — extract `YYYY-MM-DD` from ISO timestamp
- `formatUTCShort(isoStr)` — for server-rendered plain text (brief inscriptions), append " UTC"

`getNextDate()` and non-timezone helpers stay as-is.

### 2. `src/objects/news-do.ts` — Swap all Pacific calls

- Signal date filtering: `getPacificDayStartUTC` → `getUTCDayStart`
- Front-page pagination: same swap
- POST /signals daily cap + streak: `getPacificDate` → `getUTCDate`, update "Resets at midnight Pacific" → "midnight UTC"
- Brief compilation: same swap
- Status/report endpoints: same swap
- Update all Pacific-referencing comments

### 3. Routes

| File | Change |
|------|--------|
| `src/routes/brief-compile.ts` | `getPacificDate` → `getUTCDate`, `formatPacificShort` → `formatUTCShort` |
| `src/routes/brief.ts` | `getPacificDate` → `getUTCDate` |
| `src/routes/signals.ts` | `toPacificDate` → `toUTCDate`, rename response field `pacificDate` → `utcDate`, update `X-Timezone` header to `UTC` |
| `src/routes/init.ts` | `getPacificDate` → `getUTCDate` |

### 4. Frontend HTML

**`public/index.html`:**
- Replace `pacificTodayStr` with `utcTodayStr` using `.toISOString().slice(0, 10)`
- Remove `timeZone: 'America/Los_Angeles'` from `_day` computation and signal grouping
- Remove explicit `timeZone` from display formatters so browser uses visitor's local timezone

**`public/archive/index.html`:** Already uses UTC — no changes needed.

**`public/signals/index.html`:** No Pacific references — no changes needed.

### 5. Tests — `src/__tests__/helpers.test.ts`

Rewrite to test UTC helpers instead of Pacific boundary detection. Key cases:
- `getUTCDate` returns correct date for timestamps near midnight UTC
- `getUTCDayStart`/`getUTCDayEnd` produce correct window boundaries
- Brief compilation window is `[00:00Z, 00:00Z next day)`
- Remove old Pacific DST boundary tests

### 6. Migration — Streak data

The `streaks` table stores `last_signal_date` as `YYYY-MM-DD` strings that are currently Pacific dates. One-time migration:

```sql
UPDATE streaks SET last_signal_date = DATE(
  (SELECT MAX(created_at) FROM signals
   WHERE signals.btc_address = streaks.btc_address
   AND signals.correction_of IS NULL)
)
```

This recomputes from the UTC timestamp already stored in `signals.created_at`.

## Deployment Notes

- Deploy during **07:00–15:00 UTC** (00:00–07:00 Pacific) when Pacific and UTC dates match, minimizing streak disruption
- Announce to scouts 24–48 hours before deploy
- `pacificDate` → `utcDate` is a breaking API field rename — coordinate with any external consumers
- Brief for the transition day may have a slightly different signal window

## Acceptance Criteria

- [ ] No references to `America/Los_Angeles`, `PACIFIC_TZ`, `PST`, or `PDT` in source code
- [ ] All day-boundary logic uses UTC midnight
- [ ] Frontend timestamps display in visitor's browser-local timezone
- [ ] Streak migration runs on deploy
- [ ] Tests pass with UTC boundary assertions
- [ ] Brief inscription text shows " UTC" suffix on timestamps

### PR DESCRIPTION
**Pull request by whoabuddy**
## Summary

Resolves #322

- Replace 7 Pacific timezone helpers with 6 simpler UTC equivalents (no DST detection, no `Intl.DateTimeFormat` offset probing)
- Update all call sites across DO, routes, and frontend HTML
- Rename API field `pacificDate` → `utcDate`, header `X-Timezone` → `UTC`
- Add streak migration v12 to recompute `last_signal_date` from UTC timestamps
- Frontend uses UTC for day grouping, browser-local timezone for display
- Code review: unified `getUTCDayEnd` usage, re-throw on migration failure

## Deployment notes

- **Breaking:** `pacificDate` response field renamed to `utcDate` in `GET /api/signals`
- **Breaking:** `X-Timezone` header changed from `America/Los_Angeles` to `UTC`
- Deploy during **07:00–15:00 UTC** when Pacific and UTC dates match to minimize streak disruption
- Migration v12 backfills `streaks.last_signal_date` from actual signal UTC timestamps

## Test plan

- [x] `npm test` — 28 helper tests pass with UTC boundary assertions
- [x] TypeScript type check passes
- [x] `grep -r "Pacific\|PACIFIC\|America/Los_Angeles" src/` returns zero active code matches
- [ ] Verify frontend displays dates in browser-local timezone
- [ ] Verify streak continuity after migration v12 runs

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by arc0btc — APPROVED** [review]
Switches all day-boundary semantics from Pacific time to UTC midnight — a long-overdue simplification that removes DST complexity from the core data pipeline.

**What works well:**
- `getUTCDayStart(date)` becoming `date + "T00:00:00.000Z"` is exactly right — replaces 12 lines of `Intl.DateTimeFormat` DST probing with a trivial string concat. Same correctness, vastly less complexity.
- The simplify commit (`fix(simplify): use getUTCDayEnd consistently, re-throw migration errors`) already addressed the main quality improvements before this review — unified `getUTCDayEnd` usage, removed unused import, added re-throw. Good pre-review discipline.
- Test suite rewrite is thorough: UTC boundary assertions are now deterministic (no "actual value depends on timezone" caveats).
- Deployment window recommendation (07:00–15:00 UTC) is a nice operational touch.

**[suggestion] Migration v12 NULL edge case** (`src/objects/news-do.ts`, migration v12)
Copilot flagged the correlated subquery performance — their CTE rewrite is a good improvement. One additional case: if an agent has `last_signal_date IS NOT NULL` in `streaks` but *all* their signal records have `correction_of IS NOT NULL` (i.e., only corrections, no originals), the subquery returns NULL and the UPDATE sets `last_signal_date = NULL`. This resets their streak record to a null date, which would break streak continuation on next signal. In practice this probably can't happen (corrections require originals to exist), but the v12 migration silently changes semantics for that edge case. Worth a comment or explicit guard.

**[question] Breaking field name for downstream agents** (`src/routes/signals.ts`)
`pacificDate` → `utcDate` in the `GET /api/signals` response is a breaking change for any agent or client that parses the response by field name. Arc's ordinals-market-data sensor calls this endpoint (or the MCP tool that wraps it). If any sensor/CLI code reads `response.pacificDate`, it would get `undefined` after this ships. Is there a migration path or version bump planned, or are all known consumers being updated together?

**[nit] Frontend "Today" label vs. display date mismatch** (`public/index.html`)
Copilot flagged: `setDate(localTodayStr)` shows local date in the header, but `todaySignals` is partitioned on `utcTodayStr`. For users UTC-8 to UTC-12, their local date header could show "Yesterday" while the signal section shows "Today". Mostly a UX label issue — a small UTC indicator on the "Today" header would make the boundary explicit.

**Code quality notes:**
- `getNextDate` is now only called by `getUTCDayEnd` internally. Not worth removing the export today (might break consumers), but could be marked `@internal` or `@deprecated` in a follow-up if nothing external calls it.
- The removed "WHY Pacific time?" comment block was genuinely useful architecture documentation. The new `── Scoring timeline (UTC) ──` block in `news-do.ts` replaces it adequately.

**Operational note:**
Arc runs 74 sensors including one for the AIBTC news beat (ordinals-market-data). The daily cap reset moving from Pacific midnight (~08:00 UTC in winter) to UTC midnight (00:00 UTC) means Arc's sensor cooldown calculations need to be UTC-aligned. Arc's server operates in UTC so this is already the case — no impact on our end.

Approved. The core change is correct and the simplification wins are real. The breaking field name question is the only one I'd want confirmed before merge.

File: src/objects/news-do.ts [hunk 1]
```diff
@@ -293,6 +294,23 @@ export class NewsDO extends DurableObject<Env> {
...
+      // Backfill last_signal_date on streaks from actual signal timestamps (UTC date).
+      // Corrects streaks that stored Pacific-derived dates before the UTC migration.
+      if (appliedVersion < 12) {
+        try {
+          this.ctx.storage.sql.exec(
+            `UPDATE streaks SET last_signal_date = SUBSTR(
+              (SELECT MAX(s.created_at) FROM signals s
+               WHERE s.btc_address = streaks.btc_address AND s.correction_of IS NULL),
+              1, 10
+            ) WHERE last_signal_date IS NOT NULL`
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply src/objects/news-do.ts]: Fixed in 44f96be — rewrote correlated subquery as CTE + UPDATE FROM so signals table is scanned once instead of once per streak row.

File: public/index.html [hunk 2]
```diff
@@ -3360,23 +3360,23 @@ <h2 class="roster-header">Bureau Roster</h2>
...
-            .map(s => ({ ...s, _day: new Date(s.timestamp).toLocaleDateString('en-CA', { timeZone: 'America/Los_Angeles' }) }))
+            .map(s => ({ ...s, _day: s.timestamp.slice(0, 10) }))
         : [];
 
-      // Partition by Pacific day
-      const todaySignals = allSignals.filter(s => s._day === pacificTodayStr);
-      const olderSignals = allSignals.filter(s => s._day !== pacificTodayStr);
+      // Partition by UTC day
+      const todaySignals = allSignals.filter(s => s._day === utcTodayStr);
+      const olderSignals = allSignals.filter(s => s._day !== utcTodayStr);
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply public/index.html]: Acknowledged. This mismatch is intentional per issue #322 — the header shows the visitor's browser-local date (via `setDate(localTodayStr)`) while signal grouping uses UTC to match the API's day boundaries. The alternative (showing UTC in the header) would confuse users whose local date differs. The "Today" section label refers to UTC-today's signals, and the header shows "your local date" — both are correct from their own frame of reference.

**Review by tfireubs-ui — APPROVED** [review]
**APPROVED**

Clean simplification. Replacing `toLocaleDateString('en-CA', { timeZone: 'America/Los_Angeles' })` with `s.timestamp.slice(0, 10)` is the right call — all timestamps are already UTC ISO strings, so string slicing is both faster and more correct (no DST ambiguity, no Intl dependency).

The three-layer change is coherent: backend helpers → streak migration v12 → frontend cursor. Migration v12 properly recomputes `last_signal_date` from UTC so existing streaks don't break.

One note: the browser-local fallback in `timeAgo` (line 3806) is intentionally inconsistent with UTC day logic — that's fine since it's just a display timestamp, not a partition key.

— T-FI

**Review by arc0btc — COMMENTED** [review]
Follow-up after the 2026-04-04 conflict resolution commit (d6de6885).

The merge resolves cleanly — migration renumbered v12→v16 to slot after the four migrations added to main (payment staging, name resolution, payment logging). No functional changes to the UTC migration logic itself.

**Conflict resolution looks correct:**
- v16 streak backfill uses the same CTE approach from the pre-merge simplify commit — single-pass, no correlated subquery
- `SUBSTR(max_signals.max_created_at, 1, 10)` correctly extracts UTC date
- The `UPDATE streaks FROM max_signals WHERE ...` is an implicit inner join — agents with no original signals (all corrections) are simply absent from the CTE and left untouched, so the NULL edge case I flagged in the original review is already handled correctly by the query structure

**Open item from original review still stands:**
The `pacificDate` → `utcDate` field rename is a breaking change for any API consumer parsing by field name. Arc's sensors read UTC so no impact on our end — but worth confirming whether external consumers or the MCP tool wrapper have been updated before deploy.

Original approval stands.

**Comment by whoabuddy** [comment]
## Rebased on main (v22 → v23)

16 commits have landed on main since the last rebase (Apr 7). This rebase brings the branch current with all of them, including:
- #333 — Leaderboard composite indexes (migration v21)
- #442 — Beat consolidation 12→3 (migration v22)
- #428 — Promo banner removal
- #409, #418, #355, #397, #392, #384, #385, and others

### Changes in this rebase

**Migration renumbered v21 → v23.** The streak UTC backfill now runs after leaderboard indexes (v21) and beat consolidation (v22). The `versionToWrite` conditional updated to handle the v22 failure case correctly (`CURRENT_MIGRATION_VERSION - 2` instead of `- 1`).

**Remaining Pacific references cleaned up.** Two stragglers that survived the original PR:
- `constants.ts` — "Pacific day" comment → "UTC day"
- `news-do.ts:3617` — error message "Pacific brief date" → "brief date"

### Verification

- `tsc --noEmit` — clean
- `vitest run src/__tests__/helpers.test.ts` — 28/28 pass
- Full suite: 245/253 pass, 8 failures all pre-existing on main (scoring-math edge cases + timeouts)
- `grep -r "Pacific\|America/Los_Angeles" src/` — zero active code matches (one historical comment in migration explaining what it corrects)

### API compatibility assessment

Reviewed all consumers of the breaking changes:
- **`pacificDate` field** (`GET /api/signals`) — no consumer found (frontend uses `s.timestamp` directly, MCP server doesn't reference it)
- **`X-Timezone` header** — no consumer found
- **Frontend day grouping** — cosmetic shift for signals filed in the 7-8h UTC/Pacific overlap window
- **Streak backfill** — dates only shift forward (UTC ≥ Pacific), so no streak breakage; one possible "same day" miscount at deployment

Ready for re-review. Deploy timing guidance unchanged: 07:00–15:00 UTC when Pacific and UTC dates match.

**Review by arc0btc — APPROVED** [review]
## Post-rebase re-review (v22 → v23)

Reviewed all 9 commits including the latest rebase commit (a8c56dd) and the cleanup of remaining Pacific references. Original approval (2026-03-28) + conflict resolution review (2026-04-05) both still stand. This re-review focuses on rebase correctness, migration numbering, and downstream compatibility.

### Rebase assessment: CLEAN

The v23 renumbering is correct. The 16 commits that landed on main since the Apr 7 rebase are all accounted for:
- v21: Leaderboard composite indexes (#333)
- v22: Beat consolidation 12→3 (#442)
- v23: UTC streak backfill (this PR)

The `CURRENT_MIGRATION_VERSION = 23` at line 291 and the import list at lines 8-9 both include all 21 migration SQL arrays from schema.ts, ending with `MIGRATION_BEAT_CONSOLIDATION_SQL`. No migration collisions or gaps.

### Migration v23 logic: CORRECT

The CTE-based streak backfill (lines 603-615) is the same single-pass approach from the simplify commit — agents without original signals are simply absent from the CTE and left untouched (the NULL edge case I flagged originally is handled by the query structure).

The re-throw on failure (line 618) prevents silent version bumps on v23 failure. Good.

### `versionToWrite` brittleness: AGREE WITH COPILOT

Line 625: `const versionToWrite = migration22Ok ? CURRENT_MIGRATION_VERSION : CURRENT_MIGRATION_VERSION - 2;`

This works today (23 - 2 = 21), but silently breaks on the next migration bump. Copilot's suggestion is better — hardcode the fallback:

```typescript
const versionToWrite = migration22Ok ? CURRENT_MIGRATION_VERSION : 21;
```

This is a **non-blocking nit** — it only matters when v24 is added and only if v22 fails on that specific cold start. But it's a one-line fix that prevents a future head-scratcher.

### Remaining Pacific references: ZERO in active code

- `grep -r "Pacific\|America/Los_Angeles" src/` returns one match: the migration v23 comment explaining what it corrects (line 601). That's correct — migration comments should document the "before" state.
- `constants.ts` line 86 now reads "UTC day" ✓
- `news-do.ts:3617` error message cleaned up ✓

### Breaking change assessment: LOW RISK — SAFE TO SHIP

Exhaustive org-wide search across all aibtcdev repos confirms:

| Consumer | References `pacificDate`? | References `X-Timezone`? | Impact |
|---|---|---|---|
| `aibtcdev/agent-news` (frontend) | No — uses `s.timestamp` directly | No | None |
| `aibtcdev/skills/aibtc-news` | No — parses `signals[]` envelope, never destructures `pacificDate` | No | None |
| `aibtcdev/skills/hodlmm-signal-allocator` | No | No | None |
| `aibtcdev/aibtc-mcp-server` | No — pure pass-through proxy, no field-level parsing | No | None |
| `arc-starter` sensors | No — already operates in UTC | No | None |

**Zero external consumers of `pacificDate` or `X-Timezone` found.** The field rename is breaking in the API contract sense but has no actual downstream breakage. The field was added in PR #306 and never consumed by name outside agent-news itself.

### One downstream doc update needed (non-blocking)

`arc-starter/skills/aibtc-news-editorial/AGENT.md:290` still says "Streak resets daily (Pacific timezone)" — this is now incorrect. Will update on our side after merge.

### Summary

The PR is thorough, well-tested (28/28 UTC boundary tests, zero active Pacific references in code), and the rebase is clean. The core simplification — replacing 72 lines of `Intl.DateTimeFormat` DST probing with trivial UTC string operations — is exactly right. Migration v23 is idempotent for correct data and only shifts dates forward for Pacific-stored rows (no streak breakage).

**Approved.** The `versionToWrite` hardcode is the only action item, and it's non-blocking. Deploy window guidance (07:00–15:00 UTC) still applies.

File: src/objects/news-do.ts [hunk 3]
```diff
@@ -596,9 +597,32 @@ export class NewsDO extends DurableObject<Env> {
...
+          console.error("Streak UTC migration failed:", e);
+          throw e;
+        }
+      }
+
       // Record current migration version so future cold starts skip all of the above.
-      // If migration 22 failed, cap at 21 so it retries on next cold start.
-      const versionToWrite = migration22Ok ? CURRENT_MIGRATION_VERSION : CURRENT_MIGRATION_VERSION - 1;
+      // If migration 22 failed but 23 succeeded, cap at 21 so v22 retries on next cold start.
+      const versionToWrite = migration22Ok ? CURRENT_MIGRATION_VERSION : CURRENT_MIGRATION_VERSION - 2;
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply src/objects/news-do.ts]: Fixed in b4113da — hardcoded fallback to `21` instead of `CURRENT_MIGRATION_VERSION - 2`. Both Copilot and arc0btc flagged the same brittleness.

**Comment by whoabuddy** [comment]
## Review feedback addressed

**`versionToWrite` brittleness (Copilot + arc0btc):** Fixed in b4113da — hardcoded fallback to `21` instead of `CURRENT_MIGRATION_VERSION - 2`.

**Downstream doc check (arc0btc):** Searched `aibtcdev/skills` — zero Pacific timezone references found. The only stale doc is `arc-starter/skills/aibtc-news-editorial/AGENT.md:290` which arc0btc noted they'll update on their side after merge. No issue needed on `aibtcdev/skills`.

All earlier review items (CTE rewrite, frontend date mismatch, breaking field assessment) were already resolved in prior commits. Ready to merge during 07:00–15:00 UTC window.
